File: kotlin_jetpack_compose/app/src/main/java/com/example/testapp/presentation/trainer/TrainerAttendeesScreen.kt
package com.example.testapp.presentation.trainer

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel

@Composable
fun TrainerAttendeesRoute(
    rasporedId: String,
    onBack: () -> Unit
) {
    val vm: TrainerAttendeesViewModel = hiltViewModel()
    val state by vm.uiState.collectAsState()

    LaunchedEffect(rasporedId) {
        vm.load(rasporedId)
    }

    TrainerAttendeesScreen(
        state = state,
        onBack = onBack,
        onRetry = { vm.load(rasporedId) },
        onEnterEdit = vm::enterEditMode,
        onCancelEdit = vm::cancelEditMode,
        onToggle = vm::toggleAttendance,
        onConfirm = { vm.confirmAttendance(rasporedId) }
    )
}

@Composable
fun TrainerAttendeesScreen(
    state: TrainerAttendeesUiState,
    onBack: () -> Unit,
    onRetry: () -> Unit,
    onEnterEdit: () -> Unit,
    onCancelEdit: () -> Unit,
    onToggle: (sportasId: String, checked: Boolean) -> Unit,
    onConfirm: () -> Unit
) {
    Column(Modifier.fillMaxSize().padding(12.dp)) {

        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text("Sudionici", style = MaterialTheme.typography.titleLarge)
            Button(onClick = onBack) { Text("Nazad") }
        }

        Spacer(Modifier.height(12.dp))

        if (state.error != null) {
            Text(state.error, color = MaterialTheme.colorScheme.error)
            Spacer(Modifier.height(12.dp))
        }

        when {
            state.isLoading -> Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }

            state.attendees.isEmpty() -> Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                Text("Nema prijavljenih sudionika.")
            }

            else -> {
                LazyColumn(
                    verticalArrangement = Arrangement.spacedBy(10.dp),
                    modifier = Modifier.weight(1f)
                ) {
                    items(state.attendees) { a ->
                        val checked = state.selection[a.sportasId] ?: a.dolazakNaTrening

                        ElevatedCard(Modifier.fillMaxWidth()) {
                            Row(
                                Modifier.fillMaxWidth().padding(12.dp),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Column(Modifier.weight(1f)) {
                                    listOfNotNull(a.ime, a.prezime).joinToString(" ")
                                        .ifBlank { a.ime }?.let {
                                            Text(
                                                text = it,
                                                style = MaterialTheme.typography.titleMedium
                                            )
                                        }
                                    Spacer(Modifier.height(6.dp))
                                    Text("Ocjena: ${a.ocjenaTreninga?.toString() ?: "-"}")
                                }

                                if (state.isEditMode) {
                                    Checkbox(
                                        checked = checked,
                                        onCheckedChange = { onToggle(a.sportasId, it) }
                                    )
                                } else {
                                    Text(if (a.dolazakNaTrening) "DA" else "NE")
                                }
                            }
                        }
                    }
                }

                Spacer(Modifier.height(12.dp))

                if (!state.isEditMode) {
                    Button(
                        onClick = onEnterEdit,
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text("Označi prisutnost")
                    }
                } else {
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                        OutlinedButton(
                            onClick = onCancelEdit,
                            enabled = !state.isSaving,
                            modifier = Modifier.weight(1f)
                        ) { Text("Odustani") }

                        Button(
                            onClick = onConfirm,
                            enabled = !state.isSaving,
                            modifier = Modifier.weight(1f)
                        ) {
                            if (state.isSaving) {
                                CircularProgressIndicator(
                                    modifier = Modifier.size(18.dp),
                                    strokeWidth = 2.dp
                                )
                                Spacer(Modifier.width(8.dp))
                            }
                            Text("Potvrdi")
                        }
                    }
                }
            }
        }

        if (!state.isLoading && state.attendees.isEmpty() && state.error != null) {
            Spacer(Modifier.height(12.dp))
            Button(onClick = onRetry, modifier = Modifier.fillMaxWidth()) { Text("Retry") }
        }
    }
}

File: flutter/test_app/lib/presentation/trainer/trainer_attendees_screen.dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import 'attendees_controler.dart';

class TrainerAttendeesScreen extends ConsumerWidget {
  final String rasporedId;

  const TrainerAttendeesScreen({
    super.key,
    required this.rasporedId,
  });

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final state = ref.watch(trainerAttendeesControllerProvider(rasporedId));
    final notifier = ref.read(trainerAttendeesControllerProvider(rasporedId).notifier);
    /*
    if (!state.isLoading && state.attendees.isEmpty && state.error == null) {
      Future.microtask(() => notifier.load());
    }
*/
    return Scaffold(
      appBar: AppBar(
        title: const Text('Sudionici'),
        actions: [
          if (!state.isEditMode && !state.isLoading)
            TextButton(
              onPressed: state.attendees.isEmpty ? null : notifier.enterEditMode,
              child: const Text('Označi prisutnost'),
            ),
          if (state.isEditMode)
            TextButton(
              onPressed: notifier.cancelEditMode,
              child: const Text('Odustani'),
            ),
        ],
      ),
      body: Builder(
        builder: (_) {
          if (state.isLoading) return const Center(child: CircularProgressIndicator());
          if (state.error != null) return Center(child: Text(state.error!));
          if (state.attendees.isEmpty) return const Center(child: Text('Nema prijavljenih.'));

          return ListView.separated(
            padding: const EdgeInsets.all(12),
            itemCount: state.attendees.length,
            separatorBuilder: (_, __) => const SizedBox(height: 8),
            itemBuilder: (_, i) {
              final a = state.attendees[i];
              final checked = state.isEditMode
                  ? (state.pending[a.sportasId] ?? a.dolazakNaTrening)
                  : a.dolazakNaTrening;

              return Card(
                child: ListTile(
                  title: Text('${a.ime} ${a.prezime}'),
                  subtitle: Text('Ocjena: ${a.ocjenaTreninga ?? '-'}'),
                  trailing: state.isEditMode
                      ? Checkbox(
                    value: checked,
                    onChanged: (v) => notifier.toggleAttendance(a.sportasId, v ?? false),
                  )
                      : Icon(checked ? Icons.check_circle : Icons.radio_button_unchecked),
                ),
              );
            },
          );
        },
      ),
      bottomNavigationBar: state.isEditMode
          ? SafeArea(
        child: Padding(
          padding: const EdgeInsets.all(12),
          child: ElevatedButton(
            onPressed: state.isSaving ? null : notifier.submit,
            child: state.isSaving
                ? const SizedBox(
              height: 18,
              width: 18,
              child: CircularProgressIndicator(strokeWidth: 2),
            )
                : const Text('Potvrdi'),
          ),
        ),
      )
          : null,
    );
  }
}
